### ISSUE TITLE
ParameterValues in QuerySqlGenerator

### ISSUE THREAD

**Issue by cincuranet** [dotnet/efcore#18927]
Can I get access to `ParameterValues` (if known, i.e. not compiled query) in `QuerySqlGenerator` as before in 2.2? I'm using this to specialize query in some cases to get better performance (and setting `IsCacheable` to `false`).

**Comment by smitpatel** [comment]
Likely not. Currently QuerySqlGenerator works without knowing about parameters. Ideally it should not need to know parameter values as it is just a printer. If by any reason, generated SQL is different based on parameter value then it should be transformed before calling into QuerySqlGenerator.

**Comment by cincuranet** [comment]
Yes, ideally it should not. For performance reasons it's sometimes nice to have it. 

What place would be good place to intercept parameter expression (where parameters are possibly available) and transform it to other expression?

**Comment by smitpatel** [comment]
@cincuranet - I do not think it is nice to have for any reason. If printing SQL is dependent on parameter values then it is wrong design. SqlGen should print out SelectExpression as is. If something needs to change based on parameter values then it should be done before passing the SelectExpression to SqlGen. https://github.com/aspnet/EntityFrameworkCore/blob/master/src/EFCore.Relational/Query/Internal/ParameterValueBasedSelectExpressionOptimizer.cs does all the optimizations based on parameter values but notice that it does not print it out. We are looking to make it provider override-able in 5.0

**Comment by cincuranet** [comment]
> I do not think it is nice to have for any reason. If printing SQL is dependent on parameter values then it is wrong design.

Yes, but there's a difference between real world a ideal cases. At the end of the day I either can claim that something is wrong with design on whatever level or have to work around it. Guess what the person facing the issue wants to hear. :)

> ParameterValueBasedSelectExpressionOptimizer.cs

Let's hope for the best.
